### ISSUE TITLE
Typed `with` handlers are unsafe

### ISSUE THREAD

**Issue by mkantor** [gvergnaud/ts-pattern#73]
**Describe the bug**
Consider this example:

```ts
const f = (x: number) => x.toLocaleString();
const g = (x: string) => x.toUpperCase();
const input = Math.random() > 0.5;
match(input).with(true, f).with(false, g).exhaustive();
```

This typechecks, but is guaranteed to crash at runtime (the handler functions do not accept `boolean`s).

**Code Sandbox with a minimal reproduction case**
https://codesandbox.io/s/wonderful-zeh-gw5r7

**Versions**
- TypeScript version: 4.4.4
- ts-pattern version: 3.3.4
- environment: all environments

**Comment by mkantor** [comment]
Here's a more minimal example:
```ts
match(1).with(1, (x: string) => x);
```

**Comment by int8array** [comment]
you shuold use when

**Comment by gvergnaud** [comment]
Thanks for the report, I really wonder how nobody noticed this before! 

I pushed this commit which fixes this type issue:  https://github.com/gvergnaud/ts-pattern/commit/2d750742d95d38e4cab5ff1f2915050f2081441a

I'll follow-up with a release shortly

**Comment by gvergnaud** [comment]
Fixed in v3.3.5: https://github.com/gvergnaud/ts-pattern/releases/tag/v3.3.5

**Comment by mkantor** [comment]
Thanks!
